demos/RigidBodyPlanningWithQuotientSpaces.cpp

```diff
@@ -0,0 +1,154 @@
+*    permission.
+*
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+
+/* Author: Andreas Orthey */
+
+#include <ompl/base/spaces/SE3StateSpace.h>
+#include <ompl/base/spaces/SE2StateSpace.h>
+#include <ompl/base/spaces/RealVectorStateSpace.h>
+#include <ompl/base/SpaceInformation.h>
+#include <ompl/base/StateSpace.h>
+#include <ompl/geometric/planners/rrt/RRTConnect.h>
+#include <ompl/geometric/planners/quotientspace/MultiQuotient.h>
+#include <ompl/geometric/planners/quotientspace/QRRT.h>
+#include <iostream>
+#include <boost/math/constants/constants.hpp>
+
+
+namespace ob = ompl::base;
+namespace og = ompl::geometric;
+
+//Path Planning in SE2 = R2 \times SO2
+//using quotient-spaces R2 and SE2
+bool isStateValid_SE2(const ob::State *state) 
+{
+    // disallow going through the middle of the square [0,1]x[0,1]
+    // to make it more interesting (i.e. an initial R2 path might be invalid)
+    const auto *SE2state = state->as<ob::SE2StateSpace::StateType>();
+    const auto *R2 = SE2state->as<ob::RealVectorStateSpace::StateType>(0);
+    const auto *SO2 = SE2state->as<ob::SO2StateSpace::StateType>(1);
+    double *pos = R2->values;
+    double pos_cnstr =
+            sqrt((pos[0] - 0.5) * (pos[0] - 0.5) + (pos[1] - 0.5) * (pos[1] - 0.5));
+    return (pos_cnstr > 0.2) && (SO2->value < boost::math::constants::pi<double>() / 2.0);
+}
+
+bool isStateValid_R2(const ob::State *state) 
+{ 
+    return true; 
+}
+
+int main(int argc, const char **argv) 
+{
+    // Setup SE2
+    ob::StateSpacePtr SE2(std::make_shared<ob::SE2StateSpace>());
+    ob::RealVectorBounds bounds(2);
+    bounds.setLow(0);
+    bounds.setHigh(1);
+    std::static_pointer_cast<ob::SE3StateSpace>(SE2)->setBounds(bounds);
+    ob::SpaceInformationPtr si_SE2(std::make_shared<ob::SpaceInformation>(SE2));
+    si_SE2->setStateValidityChecker(isStateValid_SE2);
+
+    // Setup Quotient-Space R2
+    ob::StateSpacePtr R2(std::make_shared<ob::RealVectorStateSpace>(2));
+    std::static_pointer_cast<ob::RealVectorStateSpace>(R2)->setBounds(0, 1);
+    ob::SpaceInformationPtr si_R2(std::make_shared<ob::SpaceInformation>(R2));
```

Comment:
Again, could use auto. When using auto in line 85, the compiler is smart enough to convert R2 to the right type here.